### PR TITLE
create_cutline expose rasterio rowcol op parameter to fix pixel raste…

### PR DESCRIPTION
**Pull request by Martenz**
create_cutline expose rasterio rowcol op parameter to fix pixel raster x,y approximation

Related issue: #758

### CODE REVIEW

**Comment by vincentsarago** [comment]
thanks @Martenz 
Would you be able to add tests 🙏

**Comment by Martenz** [comment]
> thanks @Martenz Would you be able to add tests 🙏

It will take some time from my side

**Comment by Martenz** [comment]
Hi, I added the operator parameter to the existing unit tests, and I also added a test to check and raise an exception if the operator is not one of the accepted methods for the rasterio.transform.rowcol function. This means the cutline is now tested against previous implementations, allowing future users to specify something other than math.floor to obtain XS, YS pixel values if needed.

As for testing my specific issue (#758), I'm not entirely sure how to proceed. I think the problem arises because the mask is the exact contour line of a group of pixels, and with the applied approximations, the default math.floor operator primarily selects the pixels on the left side, with a few pixels inside the mask. This happens regardless of any adjustments to vrt_options. Using the round operator seems to be the only solution for accurate pixel selection in this case.

To test this behavior, I might add a fixture with a cog.tif file and a small feature mask that matches the exact pixel boundaries. However, I'm unsure about what exactly to assert in this case.

Anyhow just to double check I run all test also replacing the operator op to all test with "round" and they pass successfully 100%. 
Would this solve a note I found in the code  **tests/test_io_rasterio.py#L764** ?

Apologies if my unit testing skills are lacking, and thank you for your support.

File: rio_tiler/utils.py [hunk 1]
```diff
@@ -671,20 +683,27 @@ def create_cutline(
         str: WKT geometry in form of `POLYGON ((x y, x y, ...)))
 
     """
+
+    # Validate that the function provided is one of the allowed methods
+    if op not in {math.floor, math.ceil, round}:
```
Comment (vincentsarago) [inline rio_tiler/utils.py:688]: I don't think we need to tests this, I feel it should be done at rasterio level

File: rio_tiler/utils.py [hunk 2]
```diff
@@ -657,6 +668,7 @@ def create_cutline(
     src_dst: Union[DatasetReader, DatasetWriter, WarpedVRT],
     geometry: Dict,
     geometry_crs: CRS = None,
+    op: Callable[[float], int] = math.floor,
```
[the remark this answers is not in the document]
Comment (vincentsarago) [reply rio_tiler/utils.py:671]: we could also simply have `Callable[[float], Union[float, int]]`

[the remark this answers is not in the document]
Comment (vincentsarago) [reply rio_tiler/utils.py:671]: we could also make it optional and do 

```
op = or or numpy.floor
```

File: rio_tiler/utils.py [hunk 3]
```diff
@@ -657,6 +670,7 @@ def create_cutline(
     src_dst: Union[DatasetReader, DatasetWriter, WarpedVRT],
     geometry: Dict,
     geometry_crs: CRS = None,
+    op: Optional[Callable[[float], Union[int, float]]] = None,
```
Comment (vincentsarago) [inline rio_tiler/utils.py:673]: I feel we could just have `Optional[Callable[[float], Any]]` because the type of `numpy.floor` won't match ☝️

**Comment by vincentsarago** [comment]
@Martenz made some edits to your PR, hope you're 👍 

I'll wait for your 👍 before merging

**Comment by Martenz** [comment]
Thanks @vincentsarago for your support on this, and sorry for my late reply. I think is fine as far as I can use the round now. I will test it tomorrow.

**Comment by Martenz** [comment]
Works great as tested. Thanks again, I hope to contribute more.

**Pull request merged**